clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -523,6 +626,21 @@ private static int parseAcks(String acksString) {
      * Implementation of asynchronously send a record to a topic.
      */
     private Future<RecordMetadata> doSend(ProducerRecord<K, V> record, Callback callback) {
+        if (transactionState != null && transactionState.isTransactional() && !transactionState.hasPid()) {
+            if (transactionState.isFenced()) {
+                throw Errors.INVALID_PRODUCER_EPOCH.exception();
+            }
+            throw new IllegalStateException("Cannot perform a 'send' before completing a call to initTransactions when transactions are enabled.");
+        }
+
+        if (transactionState != null && transactionState.isCompletingTransaction()) {
+            throw new IllegalStateException("Cannot call send while a commit or abort is in progress.");
+        }
+
+        if (transactionState != null && transactionState.isInTransaction()) {
+            transactionState.maybeAddPartitionToTransaction(new TopicPartition(record.topic(), record.partition()));
```

Comment:
I think we need to move this check below since we haven't assigned the partition yet. We should probably add a test case for this.